src/main/java/formflow/library/config/SubflowManager.java

```diff
@@ -0,0 +1,201 @@
+        this.flowConfigurations = flowConfigurations;
+    }
+
+    public boolean subflowHasRelationship(String flow, String subflow) {
+        FlowConfiguration flowConfiguration = getFlowConfiguration(flow);
+        if (flowConfiguration.getSubflows().containsKey(subflow)) {
+            return flowConfiguration.getSubflows().get(subflow).getRelationship() != null;
+        } else {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND,
+                    String.format("Subflow %s not found in flow %s. Check that your flows-config.yaml is configured correctly.", subflow, flow));
+        }
+    }
+    
+    public void addSubflowRelationshipData(ScreenNavigationConfiguration currentScreen, String flow, Submission submission) {
+        String subflowName = currentScreen.getSubflow();
+        if (!subflowHasRelationship(flow, subflowName)) {
+            // No relationship so we don't need to do anything
+            return;
+        }
+
+        SubflowConfiguration currentSubflow = getSubflowConfiguration(flow, subflowName);
+        String relatedSubflowName = currentSubflow.getRelationship().getRelatesTo();
+        String relatedIdKey = currentSubflow.getRelationship().getRelationAlias();
+
+        List<Map<String, Object>> relatedSubflowData = getSubflowData(submission, relatedSubflowName);
+        List<Map<String, Object>> currentSubflowData = getOrCreateSubflowData(submission, subflowName);
+
+        if (!submission.getInputData().containsKey(subflowName)) {
+            // Initial setup: add all related items as incomplete iterations
+            relatedSubflowData.forEach(relatedItem ->
+                    currentSubflowData.add(createSubflowIterationWithRelationship(relatedIdKey, relatedItem.get("uuid")))
+            );
+        } else if (currentSubflowData.size() != relatedSubflowData.size()) {
+            // Reconciliation: we must have deleted some iterations, we need to reset the relationships and iteration statuses
+            // Collect the existing iteration IDs and then loop over the related subflow to find the missing iteration IDs
+            Set<Object> existingRelationIds = currentSubflowData.stream()
+                    .map(entry -> entry.get(relatedIdKey))
+                    .collect(Collectors.toSet());
+
+            relatedSubflowData.stream()
+                    .map(item -> item.get("uuid"))
+                    .filter(uuid -> !existingRelationIds.contains(uuid))
+                    .forEach(missingUuid ->
+                            currentSubflowData.add(createSubflowIterationWithRelationship(relatedIdKey, missingUuid))
+                    );
+        }
+    }
+
+    public boolean hasFinishedAllSubflowIterations(String currentSubflowName, Submission submission) {
+        List<Map<String, Object>> currentSubflowData = (List<Map<String, Object>>) submission.getInputData().get(currentSubflowName);
+
+        return currentSubflowData.stream()
```

Comment:
To me, it makes more sense to check in here: that currentSubflowData.size() != relatedSubflowData.size() when you filter by iteration is complete.

Here, you could catch any relationships that have missing data loop the user through that relationship

Reply:
I'm not sure I understand what you're asking / it looks like your message got cut off? Let's talk about it when we pair?